libfreerdp/core/settings.c

```diff
@@ -277,6 +277,151 @@ void freerdp_settings_print_warnings(const rdpSettings* settings)
+		{
+			WLog_ERR(
+			    TAG,
+			    "Mulitimonitor mode requested, but local layout has gaps or overlapping areas!");
+			WLog_ERR(TAG, "Please reconfigure your local monitor setup so that thre are no gaps or "
+			              "overlapping areas!");
+			return TRUE;
+		}
+	}
+	return FALSE;
+}
+
+static BOOL monitor_has_gaps(const rdpSettings* settings, UINT32 start, UINT32 count,
+                             const rdpMonitor* compare)
+{
+	BOOL hasNeighbor = FALSE;
+	for (UINT32 x = 0; x < count; x++)
+	{
+		if (x == start)
+			continue;
+
+		const rdpMonitor* monitor =
+		    freerdp_settings_get_pointer_array(settings, FreeRDP_MonitorDefArray, x);
+
+		if (compare->x == monitor->x + monitor->width)
+			hasNeighbor = TRUE;
+		if (compare->y == monitor->y + monitor->height)
+			hasNeighbor = TRUE;
+		if (compare->x + compare->width == monitor->x)
+			hasNeighbor = TRUE;
+		if (compare->y + compare->height == monitor->y)
+			hasNeighbor = TRUE;
+	}
+	if (!hasNeighbor)
+	{
+		WLog_ERR(TAG,
+		         "Monitor configuration has gaps! Monitor %" PRIu32 " does not have any neighbor",
+		         start);
+	}
+	return !hasNeighbor;
+}
+
+static BOOL freerdp_settings_client_monitors_have_gaps(const rdpSettings* settings)
+{
+	const UINT32 count = freerdp_settings_get_uint32(settings, FreeRDP_MonitorCount);
+	if (count <= 1)
+		return FALSE;
+
+	for (UINT32 x = 0; x < count; x++)
+	{
+		const rdpMonitor* monitor =
+		    freerdp_settings_get_pointer_array(settings, FreeRDP_MonitorDefArray, x);
+		if (monitor_has_gaps(settings, x, count, monitor))
+			return TRUE;
+	}
+	return FALSE;
+}
+
+static BOOL freerdp_settings_client_monitors_overlap(const rdpSettings* settings)
+{
+	const UINT32 count = freerdp_settings_get_uint32(settings, FreeRDP_MonitorCount);
+	for (UINT32 x = 0; x < count; x++)
+	{
+		const rdpMonitor* monitor =
+		    freerdp_settings_get_pointer_array(settings, FreeRDP_MonitorDefArray, x);
+		if (monitor_operlaps(settings, x + 1, count, monitor))
+			return TRUE;
+	}
+	return FALSE;
+}
+
+static BOOL freerdp_settings_client_monitors_check_primary_and_origin(const rdpSettings* settings)
+{
+	const UINT32 count = freerdp_settings_get_uint32(settings, FreeRDP_MonitorCount);
+	BOOL havePrimary = FALSE;
+	BOOL foundOrigin = FALSE;
+	BOOL rc = TRUE;
+	for (UINT32 x = 0; x < count; x++)
+	{
+		const rdpMonitor* monitor =
+		    freerdp_settings_get_pointer_array(settings, FreeRDP_MonitorDefArray, x);
+		if (monitor->is_primary)
+		{
+			if (havePrimary)
+			{
+				WLog_ERR(TAG, "Monitor configuration contains multiple primary monitors!");
+				rc = FALSE;
+			}
+			havePrimary = TRUE;
+		}
+
+		if (x == 0)
+		{
+			if ((monitor->x == 0) && (monitor->y == 0))
+				foundOrigin = TRUE;
+		}
+	}
+
+	if (!havePrimary)
+	{
+		WLog_ERR(TAG, "Monitor configuration does not contain a primary monitor!");
+		rc = FALSE;
+	}
+	if (!foundOrigin)
+	{
+		WLog_ERR(TAG, "Monitor configuration must start at 0/0 for first monitor!");
```

Comment:
I had in mind that the primary monitor must be in `(0,0)` but not necessarily that it is supposed to be the first of the list. IIRC I've seen windows configurations with 3 monitors organized horizontally, with the primary in the middle, and so the first of the list is not at origin, it has a negative x coordinate.

Reply:
That seems confirmed in MS-RDPBCGR at 2.2.1.3.6.1. Perhaps we could add a check for the 200x200 minimum size of the virtual desktop as specified in 2.2.1.3.6

Reply:
will do that.

Reply:
ok, fixed.